Weird bool behaviour when not defined as a flag parameter

I've been (incorrectly) using `bool` arguments without any option under the assumption that the mandatory value is actually parsed and turned into a proper `bool` according to what was passed on the CLI but I just found out that it is always `true`. This makes `bool` parameters which are not flags rather useless; IMHO they either need to be really parsed or there should be a compiler error saying: `You've been trying to use a bool parameter which is not a flag and hence will be always true after evaluation which is probably not what you want`.

@CreepySkeleton 

... but it does not work:
```
error[E0277]: the trait bound `bool: std::convert::From<&str>` is not satisfied
   --> src/main.rs:167:20
    |
167 |         send_resp: Flag,
    |                    ^^^^ the trait `std::convert::From<&str>` is not implemented for `bool`
    |
    = note: required by `std::convert::From::from`

error[E0277]: the trait bound `bool: std::convert::From<&str>` is not satisfied
   --> src/main.rs:166:27
    |
166 |         #[structopt(parse(from_str))]
    |                           ^^^^^^^^ the trait `std::convert::From<&str>` is not implemented for `bool`
    |
    = note: required by `std::convert::From::from`
```

--- a/examples/README.md
+++ b/examples/README.md
@@ -72,3 +72,7 @@ How to use `#[structopt(skip)]`.
 ### [Aliases](subcommand_aliases.rs)
 
 How to use aliases
+
+### [`true` or `false`](true_or_false.rs)
+
+How to express "`"true"` or `"false"` argument.

--- a/examples/true_or_false.rs
+++ b/examples/true_or_false.rs
@@ -1,0 +1,41 @@
+//! How to parse `--foo=true --bar=false` and turn them into bool.
+
+use structopt::StructOpt;
+
+fn true_or_false(s: &str) -> Result<bool, &'static str> {
+    match s {
+        "true" => Ok(true),
+        "false" => Ok(false),
+        _ => Err("expected `true` or `false`"),
+    }
+}
+
+#[derive(StructOpt, Debug, PartialEq)]
+struct Opt {
+    // Default parser for `try_from_str` is FromStr::from_str.
+    // `impl FromStr for bool` parses `true` or `false` so this
+    // works as expected.
+    #[structopt(long, parse(try_from_str))]
+    foo: bool,
+
+    // Of course, this could be done with an explicit parser function.
+    #[structopt(long, parse(try_from_str = true_or_false))]
+    bar: bool,
+
+    // `bool` can be positional only with explicit `parse(...)` annotation
+    #[structopt(long, parse(try_from_str))]
+    boom: bool,
+}
+
+fn main() {
+    assert_eq!(
+        Opt::from_iter(&["test", "--foo=true", "--bar=false", "true"]),
+        Opt {
+            foo: true,
+            bar: false,
+            boom: true
+        }
+    );
+    // no beauty, only truth and falseness
+    assert!(Opt::from_iter_safe(&["test", "--foo=beauty"]).is_err());
+}

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -302,6 +302,22 @@
 //! If you would like to use a custom string parser other than `FromStr`, see
 //! the [same titled section](#custom-string-parsers) below.
 //!
+//! **Note:**
+//! _________________
+//! Pay attention that *only literal occurrence* of this types is special, for example
+//! `Option<T>` is special while `::std::option::Option<T>` is not.
+//!
+//! If you need to avoid special casing you can make a `type` alias and
+//! use it in place of the said type.
+//! _________________
+//!
+//! **Note:**
+//! _________________
+//! `bool` cannot be used as positional argument unless you provide an explicit parser.
+//! If you need a positional bool, for example to parse `true` or `false`, you must
+//! annotate the field with explicit [`#[structopt(parse(...))]`](#custom-string-parsers).
+//! _________________
+//!
 //! Thus, the `speed` argument is generated as:
 //!
 //! ```

--- a/structopt-derive/src/attrs.rs
+++ b/structopt-derive/src/attrs.rs
@@ -479,6 +479,15 @@ impl Attrs {
 
                 match *ty {
                     Ty::Bool => {
+                        if res.is_positional() && !res.has_custom_parser {
+                            abort!(ty.span(),
+                                "`bool` cannot be used as positional parameter with default parser";
+                                help = "if you want to create a flag add `long` or `short`";
+                                help = "If you really want a boolean parameter \
+                                    add an explicit parser, for example `parse(try_from_str)`";
+                                note = "see also https://github.com/TeXitoi/structopt/tree/master/examples/true_or_false.rs";
+                            )
+                        }
                         if let Some(m) = res.find_method("default_value") {
                             abort!(m.name.span(), "default_value is meaningless for bool")
                         }

--- a/tests/ui/positional_bool.rs
+++ b/tests/ui/positional_bool.rs
@@ -1,0 +1,10 @@
+use structopt::StructOpt;
+
+#[derive(StructOpt, Debug)]
+struct Opt {
+    verbose: bool,
+}
+
+fn main() {
+    Opt::from_args();
+}

--- a/tests/ui/positional_bool.stderr
+++ b/tests/ui/positional_bool.stderr
@@ -1,0 +1,10 @@
+error: `bool` cannot be used as positional parameter with default parser
+
+  = help: if you want to create a flag add `long` or `short`
+  = help: If you really want a boolean parameter add an explicit parser, for example `parse(try_from_str)`
+  = note: see also https://github.com/TeXitoi/structopt/tree/master/examples/true_or_false.rs
+
+ --> $DIR/positional_bool.rs:5:14
+  |
+5 |     verbose: bool,
+  |              ^^^^

